### ISSUE TITLE
conditionalReboot issues

### ISSUE THREAD

**Issue by rippiedoos** [rackerlabs/auter#157]
First of all, I'm still fan of the conditinalReboot-script created in #152 and #144.

There is still one bug in the script, it cannot find the binary lsof on my system (OracleLinux/CentOS/RedHat 7) as the lsof-binary resides in /usr/sbin and that directory isn't in the crontab-path. So either the PATH should be temporarily fixed to include /usr/sbin or the lsof-binary should be called directly from /usr/sbin/lsof.

But at the end of the script it complains about the reboot-call:
```
May 30 09:09:13 <server> auter: INFO: Running Post-Apply script /etc/auter/post-apply.d/conditional-reboot.sh
May 30 09:09:14 <server> auter: /etc/auter/post-apply.d/conditional-reboot.sh assessed that the server needs to be rebooted. The assessments that triggered this requirement are:
May 30 09:09:14 <server> auter: Rebooting because detected deleted libraries
May 30 09:09:14 <server> auter: Rebooting because package kernel-uek-4.1.12-124.15.2.el7uek.x86_64 was updated and is in the /etc/auter/post-apply.d/conditional-reboot.sh APPLIST config
May 30 09:09:14 <server> auter: Rebooting server
May 30 09:09:14 <server> auter: ERROR: auter is already running or /var/run/auter/auter.pid exists.
May 30 09:09:14 <server> auter: ERROR: Post-Apply script /etc/auter/post-apply.d/conditional-reboot.sh exited with non-zero exit code 6. Aborting auter run.
```

And that's correct because auter is still running the post-script and it want's to run itself (again) with the reboot-call. Maybe it's a good idea to just call the reboot-function from the conditional reboot script but that would make it harder to debug it of course if ran standalone and not part of auter itself.

**Comment by rhodesn** [comment]
Just throwing some ideas out there for `lsof`:
`LIBCHECK=$($(which lsof) | grep lib | grep DEL)`
or
`LIBCHECK=$(PATH=/usr/sbin:/usr/local/sbin:$PATH lsof | grep lib | grep DEL)`

For the reboot:
`at "now +2 minutes" <<< "auter --reboot"`

Edit..
Figured I'd throw something together #158.

**Comment by rippiedoos** [comment]
Regarding the lsof-check, the which-solution doesn't work as which is depending on PATH, that doesn't include /usr/sbin. So the second version is nice! Also because on my ubuntu-machine, lsof indeed resides in /usr/bin, where on RPM-machine it resides in /usr/sbin. 

Regarding the reboot-issue, wouldn't it be nice to let the auter-process just wait until another instance of auter is finished? That way you could (manually) schedule the auter --apply and the auter --reboot (as suggested in the crontab-example) with one minute apart and the second auter-process would just wait until the first would finish and then run. In that case you could just run the reboot-program in the background without worrying about the 2 minutes like this:
`nohup auter --reboot`

Example of the scheduling-snippet:
```
PID=$(cat ${PIDFILE})
while [ -d "/proc/${PID}" ] ; do
    logit "Waiting for current auter-process with pid ${PID} to finish..."
    sleep 1
done
```

You could only wait on when the --reboot argument has been given or simply introduce a --wait flag or parameter to configure the waiting-part. The wait-argument would be my suggestion as it would be backwards compatible and can be introduced without breaking anything else.

**Comment by rhodesn** [comment]
If you're going to schedule the `auter --apply` and `auter --reboot` 1 minute apart then you might as well schedule `auter --apply --reboot` which would achieve the same thing? Unless I'm misunderstanding something.

I popped a similar update in the contrib script PR which I agree would be nicer than the `at` job. I might adjust the PR.

```sh
(while test -f $PIDFILE ; do sleep 1; done; auter --reboot)&
```

**Comment by reaperzn** [comment]
Hi @rippiedoos 
Thanks for raising this. I do agree with @nickrhodes-gh about the wait period. If the server needs to be rebooted straight after apply stage then there is the option that @nickrhodes-gh mentioned:
```auter --apply --reboot```
Or better still, you could set the AUTOREBOOT option to "yes" in /etc/auter.conf

The issue with scheduling an at job is that there may be other post-apply scripts which may take longer than 4 minutes or any set amount of time. (Thats the current proposed solution int the PR #158 )

I think the suggested command in the previous comment seems valid(ish). My only concern there could be a situation where a post-apply script is executed after conditional-reboot.sh which may fail. In that case, the server will be rebooted and may not be expected. 

If we implement the ```(while test -f $PIDFILE ; do sleep 5; done; auter --reboot) &``` command in the script, I think a reasonable measure to take here would be to rename the script from ```conditional-reboot.sh``` to ```99-conditional-reboot.sh``` and also add documentation stating that this should be the last script to be executed in the list of post-apply scripts.

**Comment by rhodesn** [comment]
Agreed that the conditional reboot script needs to be run last by auter so +1 to renaming it.

**Comment by rippiedoos** [comment]
Good idea about the sorting, but if the sorting of scripts is done alphabetically you should sort the scripts explicitly. It is nog guaranteed that bash will sort them alphabetically the way you think it does. With an explicit version-sort it a file starting with zz comes last, after files starting with numbers. Simple test-case:

```
#!/bin/bash
for file in $(ls -v *.txt); do
  echo "${file}" 
done
```
Output:
```
[root@server blaat]#./test.sh
00-test.txt
01-test.txt
99-xtest.txt
ZZ-test.txt
aa-test.txt
zz-test.txt
```
If you leave out the sorting by `ls`:
```
#!/bin/bash
for file in *.txt; do
  echo "${file}" 
done
```
 you end up with:
```
[root@server blaat]#./test2.sh
00-test.txt
01-test.txt
99-xtest.txt
aa-test.txt
zz-test.txt
ZZ-test.txt
```

Note the difference in alphabetical (lower and upper case) and numerical sorting. So sorting: good idea, but explicitly.

**Comment by nickrhodes-gh** [comment]
Auter uses bash pathname expansion which sorts alphabetically.

```sh
$ cat test.sh
#!/bin/bash
for file in ./temp/*; do
  echo "${file}"
done

$ ./test.sh
./temp/00-test.txt
./temp/01-test.txt
./temp/99-xtest.txt
./temp/aa-test.txt
./temp/zz-test.txt
./temp/ZZ-test.txt
```

**Comment by rhodesn** [comment]
The case-insensitive sorting seems a bit more intuitive to me.

**Comment by rippiedoos** [comment]
Yes, but the point being is that 99-\<scriptname\> isn't the best name, but should be something like ZZ-\<script-name\>. And in that case is zz-\<script-name\> (lower case prefix) more intuitive for my taste.

**Comment by reaperzn** [comment]
@rippiedoos This is a fairly common practice in linux. A lot of systems use this method including init, apache, cron and many more. The numbering system is quite well known but you do raise a valid point. Although this is also the case for all other cases where this is done. It will be down to the admin to understand that ordering system.

However, this is a possible (hacky and really ugly) workaround to ensure that all scripts were executed successfully:
```
(
TIMEOUT=300
TIMEOUTTEST=0
SYSLOG="/var/log/messages"
while ! diff <(awk -F"/" -v AUTER_START="$(grep -h "/usr/bin/auter.*--apply.*" "${SYSLOG}" | tail -n1)" '$0 ~ AUTER_START,0 { if ($0 ~ /INFO.*post-apply/) print $NF}' "${SYSLOG}" | sort) <(ls -1 "${POSTAPPLYSCRIPTDIR}" | sort) &>/dev/null || test -f "${PIDFILE}"; do
  [[ "${TIMEOUTTEST}" -ge "${TIMEOUT}" ]] && FAILED=1 && break
  TIMEOUTTEST=$[${TIMEOUTTEST}+5]
  sleep 5
done
if [[ "${FAILED}" -eq 1 ]]; then
  logger "$0 FAILED. Reboot required but not all auter post scripts ran. Server will not auto-reboot. Investigation required"
else
  logger "$0 completed. Rebooting server with auter"
  auter --reboot
fi
) &
```
Basically it will check the system log for all post-apply scripts which have been executed, compare them to the directory listing of /etc/auter/post-apply.d/ and if they match then execute the auter --reboot

**Comment by reaperzn** [comment]
@nrhodes91 What are your thoughts with the above code block? Any obvious issues with this that I am missing?

**Comment by rippiedoos** [comment]
The part about sorting the post-scripts-sorting is ok. And I’m aware of that a lot of other systems use this kind of sorting, but that isn’t clear at the moment in the documentation. If this sort of sorting is applied or intended in use, the documentation should be updated in the contrib-folder to reflect that. And maybe you should also skip all scripts that aren’t beginning with 2 digits, to be sure. But that kind of breaks current functionality but not doing so breaks it as well. Scripts without two digits will be run before those that start with letters. 

As for the script-block, for my taste it’s too complicated. It is a really long one liner. And if you update the documentation to reflect the script sorting and maybe build in some warnings about script-sorting that should be fine.

**Comment by reaperzn** [comment]
@rippiedoos I agree with your point about the documentation.

As for the code block... That's why I mentioned 
> this is a possible (hacky and really ugly) workaround

I agree that this is an ugly one-liner that while functional, is quite hacky.

We did discuss the idea of implementing a requirement of having 2 digits at the start of a valid script but as you said, this will break the current functionality.

I think the way forward is to update the documentation stating that all scripts need to be in the format of ```[0-9][0-9]-scriptname```. We might implement this as a warning only for the next release then at a later date we can enforce it in the run_script code.

**Comment by rippiedoos** [comment]
I'm looking forward to reviewing the documentation. Thanks for all the work again.

**Comment by reaperzn** [comment]
@nrhodes91 @rippiedoos I have merged the script changes and the file rename into the develop branch. I will leave this issue open pending the documentation update which will need to be a new PR

**Comment by rhodesn** [comment]
Added some documentation in #161.

**Comment by reaperzn** [comment]
This issue has now been resolved
